test/test_irods_http_api.py

```diff
@@ -3339,6 +3339,238 @@ def test_rodsadmin_can_invoke_truncate_on_other_users_data_objects_using_the_adm
+            self.assertEqual(r.json()['irods_response']['status_code'], 0)
+
+            # Show that the replicas contain the same data.
+            r = requests.get(self.url_endpoint, headers=rodsuser_headers, params={
+                'op': 'read',
+                'lpath': data_object,
+                'resource': 'demoResc'
+            })
+            self.logger.debug(r.content)
+            self.assertEqual(r.status_code, 200)
+            replica_0_data = r.content.decode('utf-8')
+
+            r = requests.get(self.url_endpoint, headers=rodsuser_headers, params={
+                'op': 'read',
+                'lpath': data_object,
+                'resource': resc_name
+            })
+            self.logger.debug(r.content)
+            self.assertEqual(r.status_code, 200)
+            self.assertEqual(r.content.decode('utf-8'), replica_0_data)
+
+        finally:
+            # Remove the data object.
+            r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+                'op': 'remove',
+                'lpath': data_object,
+                'catalog-only': 0,
+                'no-trash': 1
+            })
+            self.logger.debug(r.content)
+
+            # Remove the resource.
+            r = requests.post(f'{self.url_base}/resources', headers=rodsadmin_headers, data={
+                'op': 'remove',
+                'name': resc_name
+            })
+            self.logger.debug(r.content)
+
+    def test_targeting_replica_using_invalid_replica_number_or_resource_results_in_an_error(self):
+        rodsuser_headers = {'Authorization': f'Bearer {self.rodsuser_bearer_token}'}
+        data_object = f'/{self.zone_name}/home/{self.rodsuser_username}/using_invalid_replica_numbers_and_resources.txt'
+
+        try:
+            # Create a data object.
+            r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+                'op': 'touch',
+                'lpath': data_object
+            })
+            self.logger.debug(r.content)
+            self.assertEqual(r.status_code, 200)
+            self.assertEqual(r.json()['irods_response']['status_code'], 0)
+
+            with self.subTest('Writing to replica using non-integer string as replica number'):
+                r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+                    'op': 'write',
+                    'lpath': data_object,
+                    'replica-number': 'invalid',
+                    'bytes': 'bad replica number'
+                })
+                self.logger.debug(r.content)
+                self.assertEqual(r.status_code, 400)
+
+            with self.subTest('Writing to replica using incorrect replica number'):
+                r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+                    'op': 'write',
+                    'lpath': data_object,
+                    'replica-number': 100,
```

Comment:
Apparently, passing -1 as the replica number results in the iRODS server skipping/ignoring it and continuing as if the user didn't provide it.

See https://github.com/irods/irods/blob/4669cd6be829ed7bc3a4c6648b7eac408839e647/server/core/src/voting.cpp#L81-L84.

I haven't confirmed this with gdb, but as soon as I changed the integer to 100, the server rejected it and the dstream object entered a bad state, as desired.

So, the docs need to be updated to state that the behavior is unspecified if a negative replica number is provided. Or, the HTTP API just disallows negative replica numbers until the iRODS server is updated.